[2단계 - 로또(수동)] 마루(송채원) 미션 제출합니다.

chawani opened the pull request:
안녕하세요! 수동 로또 생성하기를 구현하여 제출합니다!
1단계 때랑 크게 달라진 부분은
1. 컨트롤러의 역할이 너무 많은 거 같아서(코드가 너무 길어져서..), view에서 값을 파싱하고 컨트롤러가 받도록 고쳤습니다
2. 이것 저것 지저분하게 있었던 Lottos의 생성자들을 정리해서 private로 두고, 정적 팩토리 메소드를 통해서 생성할 수 있도록 바꾸었습니다

그리고 정적 팩토리 메소드 관련해서 궁금한 점이 생겼습니다..! 네이밍 컨벤션을 봤는데 
```
from : 하나의 매개 변수를 받아서 객체를 생성
of : 여러개의 매개 변수를 받아서 객체를 생성....
```
이런 식으로 있더라구요. 정적 팩토리 메소드는 이름을 가질 수 있다는 장점이 있다고 했는데 저런 네이밍 컨벤션을 지키면 이름을 가진다는 느낌이 별로 안드는 거 같아서 의문이 생겼습니다😥
어느 상황에서 저 형식을 지켜야하는지, 아니면 createWithNumber(int number) 이런식으로 명명한 것들이랑 섞어 써도 되는건지 궁금합니다. 그리고 아무리 찾아봐도 네이밍 컨벤션에 맞는 상황을 구분하기가 어렵습니다😭 어떤 상황에 of라고 써야하는지, getInstance라고 써야하는지 이런것들을 명확히 구분할 수 있는 쉬운 예시가 있을까요..?

이번 리뷰도 잘부탁드립니다😊

chawani committed:
test: 사용하지 않는 import 삭제 및 값을 비교하는 부분 수정

chawani committed:
feat: 수동 로또를 구매하고, 남은 돈으로 자동 로또를 구매하는 기능 추가

chawani committed:
refactor: LottoController 리팩터링

- 입력 값에 대한 유효성 검사를 view로 이동
- 수동 구매 및 자동 구매를 하면서 지불 금액이 0이 되어 수익률 오류가 나던 부분을 수정
- 코드 중복 최소화

chawani committed:
refactor: Lottos가 생성자 대신 정적 팩토리 메소드를 사용하도록 수정

chawani committed:
refactor: getter로 반환되는 컬렉션들을 unmodifiable 하도록 수정

- Lotto에서 전체 로또 숫자들을 static으로 가지고 있도록 수정

toneyparky commented on src/main/java/lotto/domain/Lottos.java:
```diff
@@ -1,29 +1,44 @@
+        Lottos lottos = new Lottos(lottoList);
+        lottos.addAutoCreatedLotto(change.getMoney() / LOTTO_PRICE);
+        return lottos;
+    }
```
마루가 남겨주신 질문에 대해서 이야기해봐요!

보다 의미가 명확한 레퍼런스를 확인하시려면 Effective java라는 책의 item 1을 보시면 좋아요. 하지만 없으실 수 있으니 아래의 글을 첨부할게요. 원문과 비슷한 내용이에요. 

https://medium.com/@logishudson0218/effective-java-%EA%B0%9D%EC%B2%B4-%EC%83%9D%EC%84%B1%EA%B3%BC-%ED%8C%8C%EA%B4%B4-01-9999e6da4f66

내용을 보시면 각 네이밍 컨벤션에 차이가 있어요.  

from은 `형변환`에 초점이 맞춰져있고 of는 여러 변수를 적합한 타입으로 `집계(aggregate)`한다고 나와있어요. 여기서 집계는 여러가지를 모아서 하나로 만든다는걸 의미해요. 마루가 사용한 newInstance는 사용시마다 `매번 새로운 인스턴스를 보장`함을 이름을 통해 드러내고 싶을 때에 사용하면 되요. 굳이 이름을 통해 드러내지 않으려면 of 나 from을 사용해도 되구요.  

조금 모호하죠? 저도 처음 공부할 때에 혼란스러웠어요. 제가 제공한 레퍼런스도 이펙티브 자바라는 책의 저자가 이렇게 하자라고 `주장한 내용이지 절대적인 규칙은 아녜요`. 그래서 실제로 사용할 때에는 위와 같은 레퍼런스를 통해 팀에서 논의한 내용을 바탕으로 개발을 해요. 저희 팀은 of와 from을 주로 사용하구요. of와 from만으로도 읽는이는 이 인자를 이용하여 특정 객체를 만들겠구나라고 이해하고 넘어갈 수 있어요. 이게 컨벤션의 힘이죠!

더 표현력 있는 이름을 사용하고 싶으시다면 `fromMoney(Money money)` 정도는 괜찮을거같아요!

toneyparky commented on src/main/java/lotto/domain/Money.java:
```diff
@@ -15,6 +15,11 @@ public void add(final int money) {
+        validateMoney(this.money - money);
+        this.money -= money;
+    }
+
```
이 부분에 대한 테스트는 없군요..!  리팩터링도 TDD로 진행할 수 있어요. 
수정하려는 부분의 테스트가 있으면 이를 기반으로 진행하면 되구요. 지금과 같이 새로운 기능을 추가하는 경우에는 TDD를 시작하듯 테스트 코드부터 작성하시면 됩니다!

이외에도 `Rank#getSortedRanks`와 같이 로직이 들어있지만 테스트하지 못한 부분이 있다면 추가해볼까요?

toneyparky commented on src/main/java/lotto/controller/LottoController.java:
```diff
@@ -19,79 +18,85 @@
     public void run() {
         Money purchaseAmount = payForLotto();
-        Lottos lottos = buyLotto(purchaseAmount);
+        Lottos lottos = buyLotto(new Money(purchaseAmount.getMoney()));
```
`purchaseAmount`는 이미 Money 타입의 변수인데 getMoney하여 다시 Money 타입으로 생성하신 이유가 있을까요?

chawani replied:
수동으로 로또 몇장을 사는 경우에, purchaseAmount를 그 수동 로또 값만큼 감소시키고 재사용하여 자동 로또를 사도록 구현하다보니 수익률 계산에서 줄어든 그 값을 그대로 사용해 잘못된 값이 나오는 오류가 있었습니다..! 그래서 저렇게 수정했었는데, 리팩터링을 통해서 저 부분의 코드는 바꾸었습니다😋

toneyparky commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -15,35 +14,35 @@
 
     private final List<LottoNumber> lotto;
 
+    static {
+        lottoNumbers = IntStream.range(MINIMUM_NUMBER, MAXIMUM_NUMBER + 1)
+            .boxed()
+            .collect(Collectors.toList());
+    }
+
```
기존에 `getTotalLottoNumbers`를 통해 호출하여 생성하던 숫자를 상수로 처리해주셨군요. 좋네요~!

toneyparky reviewed (changes requested):
마루! 반가워요 😉

2 단계 미션을 적은 변화량으로 요구사항에 맞게 잘 처리해주셨네요! 1단계에서 잘 개발해주신 덕이에요!

질문주신 내용에 대한 답변과 몇 가지 궁금증을 남겨봤어요. 답변 및 반영 부탁드릴게요!

두 번째 미션 머지를 향해 조금만 더 힘내자구요!

chawani committed:
refactor: Profit 클래스 삭제

- 수익률 계산의 책임을 LottoResult가 가지도록 수정

chawani committed:
refactor: 정적 팩토리 메서드 리네이밍

chawani committed:
test: 테스트 코드 추가

chawani committed:
refactor: Rank를 반환하는 정적 팩토리 메서드 리네이밍

chawani committed:
refactor: controller 리팩터링

- 지불 관련 변수 및 메서드 명 변경
- 메서드 파라미터 최대한 축소

chawani committed:
refactor: 도메인 끼리 getter를 최대한 사용하지 않도록 기능 책임 이동

- 로또를 비교하여 Rank를 구하는 기능을 WinningLotto로 이동
- Money 클래스에서 Money 객체를 인자로 받아서 계산을 수행하도록 수정
- 총 수익을 계산하는 기능을 Rank로 이동
- 프로덕션 코드, 테스트 코드 리네이밍

chawani committed:
refactor: 수익률 계산을 Money가 하도록 수정

chawani committed:
refactor: 로또들의 Rank 매칭 수를 세서 반환하는 기능을 Lottos가 하도록 수정

chawani commented:
토니 안녕하세요~🤗
불변에 꽂혔다가 갑자기 getter 덜 쓰기에 꽂혀버려서 리팩터링하면서 많은 부분이 바뀌었네요😂
도메인끼리는 최대한 getter로 값을 가져와서 연산하기 보다는 값을 던져서 연산할 수 있도록 바꿔보았어요..!
이런식으로 리팩터링하면서 두가지 생각이 들었는데요,
1. 값을 전달해서 스스로 계산하게 하니까 좀 더 각 책임에 맞게 기능들이 정리되고 있는거 같다는 생각
2. 이 기능을 여기서 수행하는게 맞나...? 하는 생각

ㅎㅎ...서로 상반되는 두가지를 모두 느껴버렸네요.. getter를 쓰고 안쓰고의 기준을 명확하게 할 수 있는 방법이 있을까요? 2번처럼 느껴지는 연산에 대해서는 그냥 원래했던대로 getter로 값을 가져온 다음 연산을 수행하는게 나을까요?

리뷰 항상 감사합니다🙇‍♀️

toneyparky commented on src/main/java/lotto/view/OutputView.java:
```diff
@@ -1,16 +1,18 @@
-    private static final String LOTTO_COUNT_MESSAGE = "%d개를 구매했습니다.\n";
+    private static final String PAYMENT_REQUEST_MESSAGE = "구입금액을 입력해 주세요.";
+    private static final String LOTTO_BY_HAND_COUNT_REQUEST_MESSAGE = "\n수동으로 구매할 로또 수를 입력해 주세요.";
+    public static final String LOTTO_NUMBERS_BY_HAND_REQUEST_MESSAGE = "\n수동으로 구매할 번호를 입력해 주세요.";
```
클래스 내부에서만 사용되는 녀석이 public이네요~

toneyparky commented on src/test/java/lotto/domain/LottosTest.java:
```diff
@@ -2,22 +2,68 @@
+    @DisplayName("로또들의 일치 개수를 확인 - 1등")
+    void match_lottos_first() {
+        assertEquals(1, lottos.countLottoRank(winningLotto).get(Rank.FIRST));
+    }
```
assertEquals 함수의 내부를 살펴볼까요?

```java
	public static void assertEquals(Object expected, Object actual) {
		AssertEquals.assertEquals(expected, actual);
	}
```

좌측 인자가 expected이고 우측은 actual이네요~

이를 고려하여 테스트 코드에서 테스트해야할 값과 비교 값을 넣어주면 좋겠어요 😬

chawani replied:
앗! 리뷰 확인하고 반대로 썼던 부분들을 고치던 중에 1단계 리뷰가 문득 스쳐지나가더라구요..!
junit jupiter와 assertJ에 대해서 그때 너무 슬쩍 보고 지나갔던거 같아서, 다시 찾아보았어요(전에 TDD 강의도 너무 주의깊게 못들었던거같구..). 두가지를 비교해보니 assertJ가 가독성도 더 높고, 메서드 체이닝도 지원돼서 유용하고, 이렇게 반대로 쓰는 실수에 대해서도 방지를 할 수 있다는 장점이 있다고 해서 테스트 코드를 모두 assertJ를 사용할 수 있도록 고쳤습니다! 테스트를 더 풍부하게 할 수 있을거같아요. 이렇게 직접 실수를 저질러야 깊게 인지가 되는 일도 있네요.🤣

toneyparky replied:
크흐 좋네요~! 실수를 통해 배우고 성장한다!

저도 자바로 개발할 때에는 assertJ를 주로 사용하는 편이에요! 경험을 통해 마루만의 논리를 담은 개발관이 생겨나네요 👍

toneyparky commented on src/main/java/lotto/domain/Rank.java:
```diff
@@ -24,13 +25,24 @@
     }
 
+    public static Money getTotalWinningPrize(Map<Rank, Integer> lottoResult) {
+        Money totalWinningPrize = new Money(0);
```
`Money`를 0으로 초기화하는 기본생성자를 만들어줘도 유용하겠네요~!

toneyparky commented on src/main/java/lotto/domain/Money.java:
```diff
@@ -10,18 +10,27 @@ public Money(final int money) {
 
-    private void validateMoney(final int purchaseAmount) {
-        if (purchaseAmount < 0) {
+    public double calculateProfitRate(Money payment) {
+        return (double)this.money / payment.money;
+    }
```
기존에 `LottoResult`에서 아래와 같이 게터를 사용하던 행위를 적절한 객체로 잘 옮기셨네요!! 💯 

~~~java
// 기존 LottoResult
    public double calculateProfitRate(Money payment) {
        return (double)getProfit().getMoney() / payment.getMoney();
    }
~~~

이렇게 객체가 맡은 책임을 인지하고 적절한 행위를 하도록 리팩터링하는 방향성 좋아요!

toneyparky commented on src/main/java/lotto/domain/WinningLotto.java:
```diff
@@ -12,17 +13,15 @@ public WinningLotto(Lotto winningLotto, LottoNumber bonusBall) {
         }
     }
+
+    public Rank getLottoRank(Lotto lotto) {
+        int matchingCount = lotto.getMatchingCount(winningLotto);
+        boolean isBonus = (matchingCount == CHECK_BONUS_COUNT) && (lotto.contains(bonusBall));
+        return Rank.of(matchingCount, isBonus);
+    }
```
기존에는 `Lotto` 객체에서 `WinningLotto`를 비교하여 로또 랭크를 알아내는 행위를 수행했죠(== 책임을 가지고 있었죠)

이제는 `WinningLotto`에게 그 책임이 옮겨졌네요!!! 지금과 같은 리팩터링은 마루가 로또 간에 비교하여 몇 등인지 알아내는 역할을 `WinningLotto`에게 부여했다고 생각할 수 있어요. 

일반 로또가 Rank라는 개념을 알고있던 이전 구현에서 `WinningLotto`가 Rank만이 개념을 알게 된 지금의 리팩터링이 더 자연스러워 보여요 👍 

책임에 맞는 적절한 리팩터링이라고 생각해요.

toneyparky commented on src/main/java/lotto/domain/LottoResult.java:
```diff
@@ -4,36 +4,18 @@
 import java.util.Map;
 
 public class LottoResult {
```
리팩터링을 거치면서 `LottoResult` 객체의 책임이 많이 줄어들었네요. `Map<Rank, Integer>` 를 감싸고 있는 자료구조 느낌이 되어버린 느낌이에요. 

이런 상황에서 도출할 수 있는건 세 가지 정도 있을 거 같아요. 

1. 현상 유지: 마루의 로또 게임 세상에서는 `LottoResult`라는 개념은 이정도로 충분하다.
2. 제거: 마루의 로또 게임 세상에서는 `LottoResult`라는 개념이 필요 없다.   
3. 책임 부여: (풍부한 `LottoResult`의 개념이 필요하다면) 마루가 이번 리팩터링을 진행하면서 `LottoResult`에 필요한 책임임에도 다른 객체에게 이양했다. 

세 가지를 고민해보면 좋겠어요. 마루가 질문주신 언제 게터를 써야하는지, 게터를 사용하지 않으려 시도했지만 이게 적절한지에 대한 고민은 객체의 책임과 연관되어 있다고 생각해요. 게터를 사용하지 말자!라는 규칙이 있지만 게터를 사용해야 할 때도 있어요. 뷰에 필요한 정보를 보내거나, 객체 간의 책임을 적절하게 나누기 위해서 다른 객체의 정보를 사용해야할 때가 있죠. 

무조건 게터를 제거해보자!라는 시도는 학습적인 측면에서 좋아요. 시간을 들여서 고민하며 이런 저런 구현을 하다보면 깊이가 느니까요 😎

이번에는 로또 게임에 존재하는 모든 객체의 책임에 대해서 정리해보고 각 객체가 적절한 책임을 가지고 있는가를 판단해볼까요? 예를 들면, `Money` 객체가 지금 하는 일(가지고 있는 메서드)을 정리해보고 이 객체가 담당하는게 맞을지를 판단해보는거에요.

이러다보면 마루의 의문이 조금은 해소될지도? 물론 어려운 작업이라 고민 많이 하실겁니다 🤓

toneyparky commented on src/main/java/lotto/domain/Lottos.java:
```diff
@@ -1,29 +1,57 @@
 
-    public Lottos(Money purchaseAmount) {
-        this(purchaseAmount.getMoney() / LOTTO_PRICE);
+    public Map<Rank, Integer> countLottoRank(WinningLotto winningLotto) {
+        Map<Rank, Integer> lottoResult = new HashMap<>();
+        Arrays.stream(Rank.values()).forEach((rank -> lottoResult.put(rank, 0)));
+        for (Lotto lotto : lottos) {
+            Rank rank = winningLotto.getLottoRank(lotto);
+            lottoResult.put(rank, lottoResult.get(rank) + 1);
+        }
+        return lottoResult;
```
로또 들의 매칭 수를 계산해서 결과를 반환하는 행위를 `LottoResult`에서 `Lottos`로 옮기셨네요. 기존 코드에서는 lottos에 대해 게터를 사용해서 로직을 진행하기 때문인걸로 보여져요. 

게터의 사용은 특정 행위가 적절한 객체에서 진행되고 있지 않다는 단서가 되죠! 그렇기에 이 부분을 잘 캐치해서 옮겨주셨다고 생각해요. 다만 이러며 `LottoResult`의 행위가 줄어들었죠. 

일부러 중복된 로직을 짜지 않는다면 한가지 행위를 특정 객체에게 맡길 수 밖에 없어요. 이를 판단하는건 마루의 머릿속에 있는 객체의 정의에 달려있구요!

toneyparky commented on src/main/java/lotto/domain/Rank.java:
```diff
@@ -24,13 +25,24 @@
+        }
+        return totalWinningPrize;
+    }
+
```
이부분에 대한 테스트가 없네요~!

`Rank` 객체가 하는 행위라면 이에 대해서 검증해줘야겠죠?!

toneyparky reviewed (changes requested):
안녕하세요 마루! 마루가 깊이 있게 고민하고 학습하는 모습이 좋네요!

이번에 반영해주신 리뷰에 코멘트와 질문을 달아봤어요. 

조금은 어려울 수 있지만 같이 고민해보죠! 즐거운 고민의 시간 되시길!

chawani committed:
refactor: 구매한 로또를 모두 출력하도록 고치며 불필요해진 메서드 삭제

- OutputView에서만 사용하는 상수의 접근제한자를 private으로 변경

chawani committed:
test: AssertJ의 assertion을 사용하도록 수정 및 테스트 코드 추가

- Lotto와 Rank의 테스트 코드 추가

chawani committed:
refactor: Rank 리팩터링

- 총 상금을 계산하는 메서드를 LottoResult로 이동 및 리네이밍
- isBonus 속성 삭제
- 랭크 반환 테스트 코드 추가

chawani committed:
refactor: Money를 long 자료형으로 변경 및 0으로 초기화 하는 생성자 생성

- LottoResult의 총 상금을 계산하는 메서드 리네이밍

chawani committed:
refactor: 자동 로또 개수 출력 오류 수정 및 상수 리네이밍

chawani commented:
토니 안녕하세요!🤗
각 클래스의 책임과 역할에 대해 고민해보고 수정할 부분은 수정하였습니다~!
로또 미션을 큰 흐름으로 나누어 보면, 

> 로또들을 당첨번호와 비교 -> 몇등인지 알아냄 -> 상금 통계 내기

이렇게 동작하게 되는데 이 흐름들만 두고 클래스들이 각각 어떤 동작을 가지고 있는지 정리해 본 결과

`Lotto` : 로또끼리 비교
`WinngLotto` : Rank와 연관되어 몇 등인지 알아냄
`Lottos` : 로또끼리 비교를 반복 수행하여 비교 결과 산출
`LottoResult` : 비교 결과를 받아서 상금 통계를 냄
`Money` : 돈 계산과 관련된 동작들을 함

각 클래스당 한가지의 기능만 하고 있는 것을 보니 현재로써는 최선의 길로 온것이 아닐까?! 생각해봅니다ㅋㅋㅋ
총 상금을 구하는 상금 통계 메서드가 원래 Rank클래스에 있었는데, Rank클래스는 그냥 '등수와 상금 정보를 저장해두었다'라는 성격이 강하다고 느껴져서 다시 LottoResult로 옮기게 되었습니다. 객체지향을 현실세계랑 엮어서 생각하는게 잘하는 일인지는 모르겠지만, 현실로 생각해봤을때 로또 등수 정보한테 가서 '나 이만큼 로또 맞췄는데 상금 계산해줘' 하는게 위화감이 든다고 느껴져서요😂 보통 현실에서는 사람이 **Rank정보**와 **내가 맞춘 로또 정보** 두가지를 가지고 계산을 해보니까 그 역할을 LottoResult에게 주는게 맞다고 생각했습니다ㅎㅎ

더 생각해보거나 고쳐야 할 부분이 있으면 알려주세요! 감사합니다😊

toneyparky commented:
수정하신 내용을 확인했는데 좋네요~! `WinningLotto`와 `Rank`는 연관이 있을 수 밖에 없죠! 

그리고 말씀주신대로 객체지향이 꼭 현실세계를 반영할 필요는 없다고 저도 생각해요. 문제를 해결하기 위한 구조를 만드는게 핵심이죠. 그 구조가 현실을 반영한다면, 반영하는거구요!

toneyparky reviewed (approved):
주말에도 열심히 리뷰 반영해주셨네요!

이전 코멘트에서 드렸던 고민거리에 대해서 깊게 생각해보시고 결정 내리셨군요! 좋아요  👍
추가 테스트 작성까지 좋습니다!

우테코 두 번째 미션은 이쯤에서 마무리 지을게요. 남은 주말은 푹 쉬시면서 다음 미션을 기다려보시길바라요!

고생하셨어요 마루 😁
